design/design-doc-composition-functions.md

```diff
@@ -175,32 +179,557 @@ metadata:
+representations of a `RunFunctionRequest` and `RunFunctionResponse`.
+
+#### Use an OCI Container
+
+In the [alpha iteration of Composition Functions][alpha-design] all Functions
+are binaries that run to completion. These processes are packaged as an OCI
+container. A sidecar container in the Crossplane pod - `xfn` - runs each of them
+in a rootless container.
+
+This doesn't scale very well, but it does offer a way to build a Function for
+folks who don't want to write code; you can get a long way just using a tool
+like https://github.com/mikefarah/yq. It also offers a way to build Functions in
+languages that don't have a 'first class' SDK available. Think of this like AWS
+Lambda, which supports 9-10 first class languages, and OCI containers as a
+catch-all for everything else.
+
+I propose we move the `xfn` Function runner out of crossplane/crossplane,
+simplify it, and offer it as an alternative way to build a Function. You could
+think of this as a special kind of Composition Function SDK that builds a
+Composition Function from an OCI container, rather than building a Function a
+literal _function_ (e.g. written in Go).
+
+![Crossplane calling two containerized Functions via gRPC](assets/design-doc-composition-functions/containerized-functions.png)
+
+Under this design each containerized Function is "wrapped" in an `xfn`-like
+Composition Function. This means each Function exists as its own Deployment.
+
+This design merges the Protobuf `RunFunctionRequest` with the YAML `FunctionIO`
+types from the alpha design, so instead of reading a `FunctionIO` from stdin a
+containerized Function would read a JSON-encoded `RunFunctionRequest`, and write
+a `RunFunctionResponse` to stdout.
+
+Building a containerized Function would work as follows.
+
+Start with a simple container that reads a JSON `RunFunctionRequest` on stdin
+and writes a JSON `RunFunctionResponse` to stdout. This one simply adds a label
+(credit to @pedjak in https://github.com/crossplane/crossplane/issues/4293) for
+the idea).
+
+```Dockerfile
+FROM alpine:3.18
+RUN apk add --no-cache jq
+ENTRYPOINT ["/bin/sh", "-c", "jq '(.desired.resources[] | .resource.metadata.labels) |= {\"labelizer.xfn.crossplane.io/processed\": \"true\"} + .' | .desired"]
+```
+
+Then build a Function from the image. We want the image to run in its own
+container, not the Function container. This allows the OCI container to run
+within its own namespaces without interfering with (or being interfered with by)
+the `xfn`-like 'adaptor'. To do this we bake the container into the Function
+artifact:
+
+```shell
+# Build your OCI image.
+$ docker build .
+Successfully built 202dc6e5df4c
+
+# Save your OCI image to a tarball.
+$ docker save | gzip > function.tgz
+
+# Add some package metadata for your Function.
+$ vim crossplane.yaml
+
+# Build a Function from the tarball.
+$ kubectl crossplane function build --from-oci-tarball=function.tgz
```

Comment:
is there a more streamlined way to do this, e.g. a single step instead of separate `docker build` and `crossplane function build` steps?

Reply:
I'm not sure. I imagine there could be, but it might involve our CLI embedding bits of Docker or shelling out to `docker`, which may not be desirable.